### ISSUE TITLE
variable action/observation spaces

### ISSUE THREAD

**Issue by nollied** [Farama-Foundation/PettingZoo#648]
does anyone here have thoughts on how to encode dynamic action/observation spaces statically [as required](https://www.pettingzoo.ml/api#additional-environment-api):
```
observation_space(agent) a function that retrieves the observation space for a particular agent. This space should never change for a particular agent ID.

action_space(agent) a function that retrieves the action space for a particular agent. This space should never change for a particular agent ID.
```

there was a useful discussion [here](https://ai.stackexchange.com/questions/9491/are-there-other-approaches-to-deal-with-variable-action-spaces).

for more context, the environment i'm designing is multi-agent with the following actions at their disposal:
- create a wallet
- send money to a wallet

the send action has 3 parameters:
- amount to send
- from wallet
- to wallet

as per the "create a wallet" action, the number of wallets to select for "to" and "from" may increase with time, that's where the dynamic-ness comes from. that and the balances of the wallets change with time.

i was thinking that i could utilize a few continuous action spaces, and mapping them to wallets/percentages of amounts available in the "from wallet". 

i'm afraid that this introduces temporal complexity that the agents might not be able to represent very well without sufficient observation features. naturally, i thought to solve this by designing good observation features, but due to the dynamic nature of the action space, this actually creates a dynamic observation space: if we have 10 wallets and the agent chooses to create another, you would want to represent 11 wallets in the next time step's observations. if i used the same encoding for the observation space as the action space, the model might not be able to make sense of the continuous observations (that are supposed to represent dynamic discrete observations).

**Comment by danlessa** [comment]
@nollied it depends on the immediate scientific goals. One general approach is to simply assume that agents have a maximum number of wallets and assume a fixed observation / action space (on which the wallet being activated is one of the obs). 

This is relatively easy to implement and tends to be numerically performant while also being a cap on the maximum simulation complexity. This is the option of choice on most multi agent sims that I see out there, specially when the research space is still wide open (generally it's more productive to start with a restricted space and relax it gradually).

Doing the more relaxed version will require more thought about how the agent learning is transferred as the action/obs space grows larger. As we can see from the SO thread, it is a long topic and sometimes is more art than science. On those situations, usually is best to keep it simple and gradually add complexity as needed

**Comment by benblack769** [comment]
For this question of selecting from a large number of wallets, I would advise trying to charachterize the wallet, rather than treating it as a discrete entity. Try to think of why you would you select one wallet over another.

If the choice of wallet can be defined as the result of a softmax over a score function where the score is the value to be obtained by choosing the wallet, then you can apply this scoring function independently on each wallet, and still get proper backpropogation. 

Just an idea.

**Comment by jjshoots** [comment]
@nollied hiya, just checking in on this issue, is it done?

**Comment by jkterry1** [comment]
Hey, I'm going to close this as I think the issue is resolved, please let us know if you need any additional help
